WalletWasabi/CoinJoin/Coordinator/Coordinator.cs

```diff
@@ -322,6 +326,43 @@ private async void Round_StatusChangedAsync(object sender, CoordinatorRoundStatu
 								}
 							}
 						}
+
+						var newIntervalTriggered = Stats.Register(round.CoinJoin);
+						// Adjust timeout: https://github.com/zkSNACKs/WalletWasabi/issues/2940
+						if (newIntervalTriggered && Stats.GetAverageWcqs().Count() > 2)
+						{
+							var lastTwo = Stats.GetAverageWcqs().TakeLast(2).ToArray();
+							var older = lastTwo[0];
+							var newer = lastTwo[1];
+
+							// If the older is better, then double down, else flip.
+							if (older > newer)
+							{
+								if (Stats.LastTimeoutAdjustment == TimeoutAdjustment.Up)
+								{
+									RoundConfig.InputRegistrationTimeout = Math.Max(60, RoundConfig.InputRegistrationTimeout * (4 / 3));
+								}
+								else
+								{
+									RoundConfig.InputRegistrationTimeout = Math.Max(60, RoundConfig.InputRegistrationTimeout * (2 / 3));
+								}
+								await RoundConfig.ToFileAsync().ConfigureAwait(false);
+							}
+							else if (newer > older)
+							{
+								if (Stats.LastTimeoutAdjustment == TimeoutAdjustment.Up)
+								{
+									RoundConfig.InputRegistrationTimeout = Math.Max(60, RoundConfig.InputRegistrationTimeout * (2 / 3));
```

Comment:
I might be wrong but check the logic again. If my last action was a good move which was Up (increase) I would do the same action to increase more.